fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java

```diff
@@ -3505,130 +3497,265 @@ public void replayRecoverPartition(RecoverInfo info) {
+            if (index.getId() != table.getBaseIndexId()) {
+                // add rollup index to partition
+                partition.createRollupIndex(index);
             }
-            AgentTaskExecutor.submit(batchTask);
+        }
+        return partition;
+    }
 
-            // estimate timeout
-            long timeout = Config.tablet_create_timeout_second * 1000L * totalTaskNum;
-            timeout = Math.min(timeout, Config.max_create_table_timeout_second * 1000);
+    private void buildPartitions(Database db, OlapTable table, List<Partition> partitions) throws DdlException {
+        if (partitions.isEmpty()) {
+            return;
+        }
+        int numAliveBackends = systemInfo.getBackendIds(true).size();
+        int numReplicas = 0;
+        for (Partition partition : partitions) {
+            numReplicas += partition.getReplicaCount();
+        }
+
+        if (partitions.size() >= 3 && numAliveBackends >= 3 && numReplicas >= numAliveBackends * 500) {
+            LOG.info("creating {} partitions of table {} concurrently", partitions.size(), table.getName());
+            buildPartitionsConcurrently(db.getId(), table, partitions, numReplicas, numAliveBackends);
+        } else if (numAliveBackends > 0) {
+            buildPartitionsSequentially(db.getId(), table, partitions, numReplicas, numAliveBackends);
+        } else {
+            throw new DdlException("no alive backend");
+        }
+    }
+
+    private int countMaxTasksPerBackend(List<CreateReplicaTask> tasks) {
+        Map<Long, Integer> tasksPerBackend = new HashMap<>();
+        for (CreateReplicaTask task : tasks) {
+            tasksPerBackend.compute(task.getBackendId(), (k, v) -> (v == null) ? 1 : v + 1);
+        }
+        return Collections.max(tasksPerBackend.values());
+    }
+
+    private void buildPartitionsSequentially(long dbId, OlapTable table, List<Partition> partitions, int numReplicas,
+                                             int numBackends) throws DdlException {
+        // Try to bundle at least 200 CreateReplicaTask's in a single AgentBatchTask.
+        // The number 200 is just an experiment value that seems to work without obvious problems, feel free to
+        // change it if you have a better choice.
+        int avgReplicasPerPartition = numReplicas / partitions.size();
+        int partitionGroupSize = Math.max(1, numBackends * 200 / Math.max(1, avgReplicasPerPartition));
+        for (int i = 0; i < partitions.size(); i += partitionGroupSize) {
+            int endIndex = Math.min(partitions.size(), i + partitionGroupSize);
+            List<CreateReplicaTask> tasks = buildCreateReplicaTasks(dbId, table, partitions.subList(i, endIndex));
+            int indexCount = partitions.get(i).getVisibleMaterializedIndicesCount();
+            int timeout = Config.tablet_create_timeout_second * countMaxTasksPerBackend(tasks);
+            int maxTimeout = (endIndex - i) * Config.max_create_table_timeout_second * indexCount;
```

Comment:
Could you please add comment for this maxTimeout to help me to understand?
And is `int` type is enough?

Reply:
updated. `int` should be big enough. the default value of  `Config. max_create_table_timeout_second` is 60, which means you can create at most 35791394 indexes here.